### PR TITLE
ARTEMIS-2216 lock-free LivePageCache + tests

### PR DESCRIPTION
**Pull request by franz1981**
Given other changes aiming to reduce contention
while dealing with paging, LivePageCacheImpl
has been reimplemented to be lock-free,
multi-producer and multi-consumer in any
of its operations.

### CODE REVIEW

**Comment by franz1981** [comment]
Please take a look before merging :+1:

**Comment by franz1981** [comment]
@michaelandrepearce I have found some time to implement a further optimization not present in the original version to greatly reduce the cost of walking the list of messages, allowing near constant cost if sequential access patterns are used on `LivePageCache::getMessage` queries.

**Comment by franz1981** [comment]
I have prepared a benchmark that shows some weird behaviour so please wait before merge this :+1:

**Comment by franz1981** [comment]
I think i have handled this one: I have found some unecessary query on the live page cache, just to compute the persistent size.
In addition, the data structure I've built was missing an important optimization of `LinkedList` ie the ability of walk backward to search an element if contained in the second half of the list.
Hence I have provided a similar (but more dinamic) way to get the same behaviour.
I wasn't expecting this change to be that big, but I have already tested that the gain is quite interesting, but I would like @qihongxu to validate my findings with a load different from mine :+1: 

@michaelandrepearce I know that you're curious about these things :P 
The chunk size I have chosen allow to be efficient if there are *at least* 9 paged message into the cache.
Each original linked list node was ~40 bytes while an `AtomicChunk` of capacity 32 is ~324 bytes (array included, 44 bytes if excluding it) so it means that for default page size (10 MB) and avg msg size of 1024 the cache need to have a capacity of 10240 slots.
10240 instances with the cache on master means 10240 nodes * 40 bytes/node = 409600 bytes
while with the PR cache means (10240/32) chunks * 324 bytes/chunk = 103680 bytes
It means a that we have stored the same number of messages in about 1/4 of the space: considering that users are expected to use paging to save JVM heap space it is a nice improvement :)

**Comment by franz1981** [comment]
I have maintained the original idea of not bloating of garbage while adding data to the cache, but obviously it can be made much faster (on `getMessage`) by using a growable data structure a-la `ArrayList` (but concurrent), the only downsides are:
- `addLiveMessage` can be blocked due to copying the data if a resize would happen
- `the GC would be triggered much more if a burst of messages would suddenly being added

**Comment by michaelandrepearce** [comment]
@franz1981

Re 

`the GC would be triggered much more if a burst of messages would suddenly being added


This is a worry as one reason a consumer may lag and cause brief period of paging is if a sudden burst occurs on producer. Have you got stats on how bad before and after?

**Comment by franz1981** [comment]
@michaelandrepearce
> the GC would be triggered much more if a burst of messages would suddenly being added

Now is always (much) better then master :)
That sentence is related to an hypothetical version that I haven't implemented yet, because I already know that is too demanding on `addLiveMessage` side from the GC side/latencies. I hope to have explained it better now :)

**Comment by michaelandrepearce** [comment]
Ah fair enough. You made it sound worse ;) .

Cool i look forward to the stats on the other pr thread then :)

**Comment by franz1981** [comment]
> Ah fair enough. You made it sound worse ;) .
> 
> Cool i look forward to the stats on the other pr thread then :)

I know, but I have spent the whole weekend playing with this data-structure (for fun too) and I was very unhappy that the most performant solution (O(1) on `getMessage`!) is the only one that cannot be used because produce too much garbage and has a very high memory footprint: a real performance vs memory trade-off :(

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageSubscriptionImpl.java [hunk 1]
```diff
@@ -461,11 +461,22 @@ public void confirmPosition(final Transaction tx, final PagePosition position) t
...
+         store.storeCursorAcknowledgeTransactional(tx.getID(), cursorId, position);
+      }
+      installTXCallback(tx, position, persistentSize);
+   }
+
    @Override
    public void ackTx(final Transaction tx, final PagedReference reference) throws Exception {
-      confirmPosition(tx, reference.getPosition());
+      //pre-calculate persistentSize
+      final long persistentSize = getPersistentSize(reference);
```
Comment (franz1981) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageSubscriptionImpl.java:475]: @clebertsuconic wdyt? It is safe to do it?

**Pull request closed without merging**